README.md
<h1 align="center"> PROYECTO SCRABBLE </h1>

# Luciana-Sosa

"ScrabbleGame is a game where 2 to 4 people can play. Each player takes turns placing words on the board, starting from the center. All words must be connected. Each player has a Rack, which is their bag of letters. With these letters, they must create words. Each letter tile has different values, which are used to determine the winner at the end. The game ends when there are no more letters left for the players. (The words are from the Spanish dictionary and will be validated by the game.) "

<!DOCTYPE html>
<html>
<head>

</head>
<body>
  <h2>Steps to run in terminals</h2>
  <div class="steps">
    <ol>
      <li>Install Docker, if you already have it installed much better!</li>
      <li>Install Git in the terminal</li>
      <li>Clone the repository using: <code>git clone https://github.com/um-computacion-tm/scrabble-2023-lgsosa.git</code></li>
      <li>Navigate to the repository: <code>cd /.../scrabble-2023-lgsosa</code></li>
      <li>Build the Docker image: <code>docker build -t image_name</code> (in the name of the image give it whatever name you want, I called it cd1a61941a13)</li>
      <li>Run Docker image: <code>docker run -it image_name</code> (In my case it would be: <code>docker run -it cd1a61941a13 </code>)</li>
    </ol>
  </div>
  <p>🏁 Proyecto Finalizado 🏁</p>
</body>
</html>

# maintainability
[![Maintainability](https://api.codeclimate.com/v1/badges/501cec3723d1db83d4e8/maintainability)](https://codeclimate.com/github/um-computacion-tm/scrabble-2023-lgsosa/maintainability)

# test coverage
[![Test Coverage](https://api.codeclimate.com/v1/badges/42bb4f6e2a7cece8f898/test_coverage)](https://codeclimate.com/github/um-computacion-tm/scrabble-2023-lgsosa/test_coverage)


CHANGELOG.md
<h1 align="center"> ScrabbleGame </h1>
by Luciana Sosa

### [0.0.35] - 2023-11-08
### Added

touch-ups

### [0.0.34] - 2023-11-07
### Added

Last revision and correction of details
Change the main because he had bad lefts and rights

### [0.0.33] - 2023-11-06
### Added

I add the last details for delivery.
I haven't uploaded my changes yet.

### [0.0.32] - 2023-11-03
### Added

Improved coverage and the player menu now looks cleaner.

### [0.0.31] - 2023-11-02
### Added

Add more tests for coverage

### [0.0.30] - 2023-11-01
### Added

I go back to fix errors in my bagtiles and my main
Changes in the board

### [0.0.29] - 2023-10-31
### Added

Changes in the test_main for coverage
Changes in the main 

### [0.0.28] - 2023-10-30
### Added

changes throughout the game (my changes are not uploaded so that they are not plagiarized)
I deleted my way of applying the dictionary, it gave me many errors

### [0.0.27] - 2023-10-29
### Added

Changes in configure_game, models and players
Changes in scrabble and configure_game

### [0.0.26] - 2023-10-28
### Added

Changes scrabble and configure_game

### [0.0.25] - 2023-10-27
### Added

Changes board, player, main, scrabble, calculate_word_value

### [0.0.24] - 2023-10-25
### Added

Changes main, get_player_count and game_complete
Changes game_complete

### [0.0.23] - 2023-10-24
### Added

Changes main, board, game_complete and tests

### [0.0.22] - 2023-10-22
### Added

Changes main, board and tests

### [0.0.21] - 2023-10-18
### Added

Changes in main, player, board, tests

### [0.0.20] - 2023-10-15
### Added

Changes in dictionary,game_board, game_calculate, game_player, main
Changes in all tests

### [0.0.19] - 2023-10-09
### Added

Changes in dictionary.py
2do commit- Changes in main.py, scrabble.py,test_main,get_player_count 

### [0.0.18] - 2023-10-08
### Added

Add dictionary

### [0.0.17] - 2023-10-02
### Added

Changes to the game presentation, in main.py

### [0.0.16] - 2023-10-01
### Added

Add test_pass_turn
Add def pass_turn
coverage 98%

### [0.0.15] - 2023-09-28
### Added

Add test_validate_user_has_letters
Coverage progress: 98%

### [0.0.14] - 2023-09-26
### Added

Add the changes in the changelog

### [0.0.13] - 2023-09-25
### Added

Add get_score   
Coverage progress: 98%

### [0.0.12] - 2023-09-20
### Added

Add main and test class
Add tests to game_board
Coverage progress: 99%

### [0.0.11] - 2023-09-16
### Added

Add the progress in the changelog.

### [0.0.10] - 2023-09-14
### Added

Add test and function of validate_word_out_board
Increases program coverage by improving features

### [0.0.9] - 2023-09-13
### Added

Add test and function of validate_word_inside_board
Add test and function of text_next_turn_when_player_is_last

### [0.0.8] - 2023-09-12
### Added

Add test and function of word-out-of-board

### [0.0.7] - 2023-09-11
### Added

I create a new branch "develop2"
Add player_last, (tests y function of next_turn)

### [0.0.6] - 2023-09-06
### Added

I add test and its corresponding function: validate_word, place_tile

### [0.0.5] - 2023-09-03
### Added

Improve the ScrabbleGame class

### [0.0.4] - 2023-08-30
### Added

I fix the branch and the CircleCi
I add all my progress so far, board class, cell class, player class, bag tiles class, scrabble game class
Add Cell classes with add_letter and calculate_value methods.

### [0.0.3] - 2023-08-29
### Added

I lost the "commits" that I advanced

### [0.0.2] - 2023-08-28
### Added

I had a bug and broke my branch

## [0.0.1] - 2023-08-21
### Added

Add Tiles and Bangtiles classes.

game/board.py
from game.cell import *
from pyrae import dle
from game.models import D as DATA
from game.player import Player

class Board:
    def __init__(self):
        self.grid = [[Cell(1, '') for _ in range(15)] for _ in range(15)]
        self.add_multipliers()

    def board_in_terminal(self):
        view = ' \n'
        view += '    0  1  2  3  4  5  6  7  8  9  10 11 12 13 14 \n'
        letters = 'ABCDEFGHIJKLMNO'
        
        for i, row in enumerate(self.grid):
            view += f"{letters[i]}   "
            for cell in row:
                if cell.tile is None:
                    multiplier = f"{cell.multiplier_type[0].upper()}{cell.multiplier}" if cell.multiplier_type else '  '
                    view += f"{multiplier}|"
                else:
                    view += f"{cell.tile:2}|"

            view += "\n"

        return view
    

    def add_multipliers(self):
        WORD_MULTIPLIERS = ((0, 0), (7, 0), (0, 7), (7, 7), (0, 14), (7, 14), (14, 0), (14, 7), (14, 14), (1, 1), (2, 2), (3, 3), (4, 4), (10, 10), (11, 11), (12, 12), (13, 13), (1, 13), (2, 12), (3, 11), (4, 10), (10, 4), (11, 3), (12, 2), (13, 1))
        LETTER_MULTIPLIERS = ((1, 5), (1, 9), (5, 1), (5, 5), (5, 9), (5, 13), (9, 1), (9, 5), (9, 9), (9, 13), (13, 5), (13, 9),(0, 3), (0, 11), (2, 6), (2, 8), (3, 0), (3, 7), (3, 14), (6, 2), (6, 6), (6, 8), (6, 12), (7, 3), (7, 11), (8, 2), (8, 6), (8, 8), (8, 12), (11, 0), (11, 7), (11, 14), (12, 6), (12, 8), (14, 3), (14, 11))
        for row in range(15):
            for col in range(15):
                cell = self.grid[row][col]
                if (row, col) in LETTER_MULTIPLIERS:
                    multiplier = 3 if (row, col) in ((1, 5), (1, 9), (5, 1), (5, 5), (5, 9), (5, 13), (9, 1), (9, 5), (9, 9), (9, 13), (13, 5), (13, 9)) else 2
                    cell.multiplier = multiplier
                    cell.multiplier_type = 'letter'
                elif (row, col) in WORD_MULTIPLIERS:
                    multiplier = 3 if (row, col) in ((0, 0), (7, 0), (0, 7), (0, 14), (7, 14), (14, 0), (14, 7), (14, 14)) else 2
                    cell.multiplier = multiplier
                    cell.multiplier_type = 'word'

    def get_cells(self, pos, i, horizontal):
        cell = self.grid[pos[0]][pos[1] + i] if horizontal else self.grid[pos[0] + i][pos[1]]
        if pos[0] > 0 and pos[1] > 0:
            invert_cell = self.grid[pos[0] - 1][pos[1] + i] if horizontal else self.grid[pos[0] + i][pos[1] - 1]
        else:
            invert_cell = None
        try:
            side_cell = self.grid[pos[0] + 1][pos[1] + i] if horizontal else self.grid[pos[0] + i][pos[1] + 1]
        except:
            side_cell = None
        return cell, invert_cell, side_cell

    def get_side_word(self, cell, i, orientation, pos, letter):
        horizontal = orientation[0]
        inverted = orientation[1]
        k = -1 if inverted else 1
        side_word = letter
        j = 1
        while cell.tile:
            side_word += cell.tile.letter
            cell = self.grid[pos[0] + (1 + j)*k][pos[1] + i] if horizontal else self.grid[pos[0] + i][pos[1] + (1 + j)*k]
            j += 1
        if inverted:
            side_word = side_word[::-1]
        return side_word, j

    def get_letter_value(self, letter):
        for tile in DATA:
            if letter == tile['letter']:
                return tile['value']
        return 0

    def update_multipliers(self, cell, letter_value, word_multiplier, points, first):
        if not (cell.multiplier_type == 'letter' ) and cell.active:
            word_multiplier *= cell.multiplier
        points += (letter_value * cell.multiplier) if cell.multiplier_type == 'letter' and cell.active else letter_value
        if first:
            cell.active = False
        return word_multiplier, points
    
    def validate_word_inside_board(self, word, location, orientation):
        if orientation == "H":
            return len(word) <= len(self.grid)-location[0]
        else:
            return len(word) <= len(self.grid)-location[1]

    def is_empty(self):
        return self.grid[7][7].tile == None
    
    def validate_empty(self, word, location, orientation):
        h_space = len(word) <= len(self.grid)-location[0]
        v_space = len(word) <= len(self.grid)-location[1]
        if orientation=='H' and h_space and location[0]==7:
            for i in range(len(word)):
                if location[1] + i == 7:
                    return True
        elif (orientation !='H' and v_space and location[1]==7):
            for i in range(len(word)):
                if location[0] + i == 7:
                    return True
        return False
    
    def validate_side_cell(self, parameters, validators, cell, index_increment):
        word, pos, i = parameters[0], parameters[1], parameters[2]
        is_valid, intersections = validators[0], validators[1]
        grid, word2, index = self.grid, word[i], 1
        while cell:
            word2 += cell
            cell = grid[pos[0] + i + index_increment * index][pos[1] + i].tile
            index += 1
        word2_is_valid = self.rae_search(word2)
        if not word2_is_valid:
            is_valid = -9999
        else:
            is_valid += 1
            intersections += 1
        return [is_valid, intersections]

    def check_cells(self, cells, parameters, validators):
        cell, sidecell, invertsidecell = cells[0], cells[1], cells[2]
        word, pos, i = parameters[0], parameters[1], parameters[2]
        is_valid, intersections = validators[0], validators[1]
        if cell:
            intersections += 1
            is_valid += 1 if cell == word[i].upper() else 0
        elif sidecell:
            result = self.validate_side_cell([word, pos, i], [is_valid, intersections], sidecell, 1)
            is_valid, intersections = result[0], result[1]
        elif invertsidecell:
            result = self.validate_side_cell([word, pos, i], [is_valid, intersections], invertsidecell, -1)
            is_valid, intersections = result[0], result[1]
        return [is_valid, intersections]
    
    def calculate_word_value(self, word, pos, orientation, first=True):
        horizontal = orientation.upper() == 'H'
        word = Player().split_word(word)
        points = 0
        word_multiplier = 1

        for i, letter in enumerate(word):
            cell, invert_cell, side_cell = self.get_cells(pos, i, horizontal)
            available = not cell.tile and i == 0

            if invert_cell and invert_cell.tile and available:
                side_word, j = self.get_side_word(invert_cell, i, (horizontal, True), pos, letter)
                points += self.calculate_word_value(side_word, (pos[0] - j + 1, pos[1] + i) if horizontal else (pos[0] + i, pos[1] - j + 1), not horizontal, False)
            elif side_cell and side_cell.tile and available:
                side_word, j = self.get_side_word(side_cell, i, (horizontal, False), pos, letter)
                points += self.calculate_word_value(side_word, (pos[0], pos[1] + i) if horizontal else (pos[0] + i, pos[1]), not horizontal, False)

            letter_value = self.get_letter_value(letter)
            word_multiplier, points = self.update_multipliers(cell, letter_value, word_multiplier, points, i == 0)

        return points * word_multiplier

    def validate_not_empty(self, word, pos, orientation):
        horizontal = True if orientation.upper() == 'H' else False
        intersections = 0
        is_valid = 0
        grid = self.grid
        for i in range(len(word)):
            cell = grid[pos[0] + (i if not horizontal else 0)][pos[1] + (i if horizontal else 0)].tile
            sidecell = grid[pos[0] + i][pos[1] + 1].tile if not horizontal else grid[pos[0] + 1][pos[1] + i].tile
            invertsidecell = grid[pos[0] + i][pos[1] - 1].tile if not horizontal else grid[pos[0] - 1][pos[1] + i].tile
            checked = self.check_cells([cell, sidecell, invertsidecell], [word, pos, i], [is_valid, intersections])
            is_valid, intersections = checked[0], checked[1]
        return is_valid != 0 and is_valid == intersections
        
    def validate(self, word, location, orientation):
        if self.validate_word(word):
            if self.grid[7][7].tile is None:
                return self.validate_empty(word, location, orientation)
            return self.validate_not_empty(word, location, orientation)
        return False
            
    def validate_side(self, word, location, orientation, direction):
        row = location[0]
        column = location[1]
        step = 1 if direction == 'right' else -1

        for i in range(len(word)):
            cell = self.grid[row + i][column]
            side_cell = self.grid[row + i][column + step]

            if cell.tile is None:
                if side_cell.tile is not None:
                    temporal_word = word[i]
                    j = 1
                    temporal_scell = side_cell

                    while temporal_scell.tile is not None:
                        temporal_word += temporal_scell.tile.letter
                        temporal_scell = self.grid[row + i][column + step + j]
                        j += 1

                    if direction == 'left':
                        temporal_word = temporal_word[::-1]

                    if self.validate_word(temporal_word):
                        return True

        return False

            
    def put_word(self,word,location,orientation):
        horizontal = True if orientation.upper() == 'H' else False
        j=0
        for i in range(len(word)):
            cell = self.grid[location[0]][location[1]+i+j] if horizontal else self.grid[location[0]+i+j][location[1]]
            while cell.tile:
                j+=1
                cell = self.grid[location[0]][location[1]+i+j] if horizontal else self.grid[location[0]+i+j][location[1]]
            cell.tile = word[i]

    def words_with_accent(self, word):
        word = word.lower()
        for letter in word:
            if letter == 'á':
                word = word.replace('á', 'a')
            elif letter == 'é':
                word = word.replace('é', 'e')
            elif letter == 'í':
                word = word.replace('í', 'i')
            elif letter == 'ó':
                word = word.replace('ó', 'o')
            elif letter == 'ú':
                word = word.replace('ú', 'u')
        return word.upper()


    def get_word_without_intersections(self,word,pos,horizontal):
        result = ''
        for i in range(len(word)):
            cell = self.grid[pos[0] + (i if not horizontal else 0)][pos[1] + (i if horizontal else 0)]
            if not cell:
                result += word[i]
        return result
    
    def validate_word(self, word):
        dle.set_log_level(log_level = 'CRITICAL')
        search = dle.search_by_word(word=word)
        if search is None:
            raise DictionaryConnectionError()
        return search.meta_description != 'Versión electrónica 23.6 del «Diccionario de la lengua española», obra lexicográfica académica por excelencia.'


class DictionaryConnectionError(Exception):
    ...